src/Razor/src/Microsoft.VisualStudio.LanguageServices.Razor/Documents/VisaulStudioFileChangeTracker.cs

```diff
@@ -58,71 +71,109 @@ public override void StartListening()
         {
             _foregroundDispatcher.AssertForegroundThread();
 
-            try
+            if (_fileChangeAdviseTask != null)
             {
-                if (_fileChangeCookie == VSConstants.VSCOOKIE_NIL)
-                {
-                    var hr = _fileChangeService.AdviseFileChange(
-                        FilePath,
-                        FileChangeFlags,
-                        this,
-                        out _fileChangeCookie);
-
-                    Marshal.ThrowExceptionForHR(hr);
-                }
+                // Already listening
+                return;
             }
-            catch (Exception exception)
+
+            if (_fileChangeUnadviseTask?.IsCompleted == false)
             {
-                _errorReporter.ReportError(exception);
+                // An unadvise operation is still processing, block the foreground thread until it completes.
+                _fileChangeUnadviseTask.Join();
             }
+
+            _fileChangeAdviseTask = _joinableTaskFactory.RunAsync(async () =>
+            {
+                if (_fileChangeUnadviseTask != null)
+                {
+                    // Let the unadvisement finish
+                    await _fileChangeUnadviseTask;
+                }
+
+                try
+                {
+                    return await _fileChangeService.AdviseFileChangeAsync(FilePath, FileChangeFlags, this);
+                }
+                catch (Exception exception) when (!(exception is PathTooLongException))
+                {
+                    // We can't prevent PathTooLongException's so we ignore those.
```

Comment:
This statement sounds like the opposite of what's happening. Everything but `PathTooLongException` gets reported, while the latter causes the task to fault?